rust/src/storage/s3/mod.rs

```diff
@@ -405,80 +453,91 @@ pub trait LockClient: Send + Sync + Debug {
     /// Attempts to acquire lock. If successful, returns the lock.
     /// Otherwise returns [`Option::None`] which is retryable action.
     /// Visit implementation docs for more details.
-    async fn try_acquire_lock(&self) -> Result<Option<LockItem>, StorageError>;
+    async fn try_acquire_lock(&self, data: &str) -> Result<Option<LockItem>, StorageError>;
 
     /// Returns current lock from DynamoDB (if any).
     async fn get_lock(&self) -> Result<Option<LockItem>, StorageError>;
 
+    /// Update data in the upstream lock of the current user still has it.
+    /// The returned lock will have a new `rnv` so it'll increase the lease duration
+    /// as this method is usually called when the work with a lock is extended.
+    async fn update_data(&self, lock: &LockItem) -> Result<LockItem, StorageError>;
+
     /// Releases the given lock if the current user still has it, returning true if the lock was
     /// successfully released, and false if someone else already stole the lock
     async fn release_lock(&self, lock: &LockItem) -> Result<bool, StorageError>;
 }
 
-#[allow(clippy::unnecessary_wraps)]
-fn try_create_lock_client(_region: Region) -> Result<Option<Box<dyn LockClient>>, StorageError> {
-    match std::env::var("AWS_S3_LOCKING_PROVIDER") {
-        Ok(p) if p.to_lowercase() == "dynamodb" => {
-            cfg_if::cfg_if! {
-                if #[cfg(feature = "dynamodb")] {
-                    let client = dynamodb_lock::DynamoDbLockClient::new(
-                        rusoto_dynamodb::DynamoDbClient::new(_region),
-                        dynamodb_lock::Options::default()
-                    );
-                    Ok(Some(Box::new(client)))
-                } else {
-                    Err(StorageError::Generic("dynamodb feature is not enabled".to_string()))
-                }
+const DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS: u32 = 10_000;
+
+impl dyn LockClient {
+    async fn rename_with_lock(
+        &self,
+        s3: &S3StorageBackend,
+        src: &str,
+        dst: &str,
+    ) -> Result<(), StorageError> {
+        let mut lock = self.acquire_lock_loop(src, dst).await?;
+
+        if let Some(ref data) = lock.data {
+            let data: LockData = serde_json::from_str(data)
+                .map_err(|_| StorageError::S3Generic("Lock data deserialize error".to_string()))?;
+
+            let mut rename_result = s3.unsafe_rename_obj(&data.source, &data.destination).await;
+
+            if rename_result.is_ok() && lock.acquired_expired_lock {
+                // If we acquired expired lock then the rename done above is
+                // a repair of expired one. So on this time we try the intended rename.
+
+                // lock.data = LockData { src, dst }
+                // s3.update_data()
+                lock.data = Some(LockData::json(src, dst)?);
+                lock = self.update_data(&lock).await?;
```

Comment:
I'm not sure I follow you, we're not reacquire a lock here, e.g. we still hold it. The `update_data` just updates current lock's data, the `rvn` is updated though because we'd like to extends `lease_duration`, so that's why new lock instance is returned. 
This design is borrowed from reference implementation of `heartbeat`, on each change they also change a `rvn` so other workers would honour the changes and extends a wait

Reply:
Yep, minor semantic difference. i consider a lock with new `rvn` a new lock in my previous comment. To me, the end result is not that different from releasing the lock, then reacquire the lock with the same data.

Let me try to add more context to the question I was asking with regards to `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS`.

Assuming no repair happened, the writer will try `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS` times in the acquire_lock_loop until it is able to acquire one. I think this semantic is very clear.

Now, let's consider the following trace:

* on the first try in `acquire_lock_loop`
* writer successfully acquired an expired lock. 
* writer completes the repair
* writer paused and lock expired
* writer resumed
* `self.update_data` failed updating lock with data

In this case, the lock update/reacquire error will be propagated back to the delta commit loop without a retry. Is this expected? In order words, do we want to make repair transparent to the caller so regardless repairs have happened or not, we always try to acquire the lock for `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS` times? If commit loop got a lock acquire error on the first try, what should it do?

Reply:
To be honest, I'd like to crash a worker if it pauses, because why should one? However, if we see it's a common behaviour and with no workaround then we can make it resilient by allowing the worker to recover from this and retry rename. 

> do we want to make repair transparent to the caller so regardless repairs have happened or not, we always try to acquire the lock for DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS times? If commit loop got a lock acquire error on the first try, what should it do?

Crash I feel like, we should recover only on the error we assured, such as condition failure (someone else holds a lock) or throughput exceeded. These two are covered in reference impl as well. 

But regarding unexpected errors (and pauses as such) I lean toward failing the rename and worker itself, since most likely the owner should take actions

Reply:
This is a fair design decision, but I think it's worth documenting somewhere in the doc string so users are aware that `DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS` won't take into affect if lock acquisition failed due to worker itself pausing for too long.

Reply:
I agree that better documenting is needed, but I'm against to adapt to the case where network is slow and outreaches the default 20 seconds. For those cases we have an option to increase a `lease_duration` to whatever values matches their needs

Reply:
yeah, I wasn't suggesting we handle network pause here, I think you decision of crash loudly on pauses is a fine design decision. just need to be documented because since there are many distributed applications designed to be able to handle pauses, so not everyone will walk in with the same assumption :)